### PR TITLE
Update MsalInterceptor to accept HTTP methods in protectedResourceMap

### PR DESCRIPTION
**Pull request by jo-arroyo**
Currently, the Msal Interceptor attaches the auth header with GET requests even if it is not needed. This PR changes the `protectedResourceMap` to allow accepting an object with HTTP methods and scopes for that method, allowing different tokens to be attached to different HTTP methods.

This PR also adds unit tests and a `MsalInterceptor` doc, and addresses issue #3246.

### CODE REVIEW

File: lib/msal-angular/src/msal.interceptor.config.ts [hunk 1]
```diff
@@ -11,6 +11,11 @@ export declare type MsalInterceptorAuthRequest = Omit<PopupRequest, "scopes"> |
 
 export type MsalInterceptorConfiguration = {
     interactionType: InteractionType.Popup | InteractionType.Redirect;
-    protectedResourceMap: Map<string, Array<string> | null>;
+    protectedResourceMap: Map<string, Array<string|ProtectedResourceScopes> | null>;
     authRequest?: MsalInterceptorAuthRequest | ((msalService: MsalService, req: HttpRequest<unknown>, originalAuthRequest: MsalInterceptorAuthRequest) => MsalInterceptorAuthRequest);
 };
+
+export type ProtectedResourceScopes = {
```
Comment (jo-arroyo) [inline lib/msal-angular/src/msal.interceptor.config.ts:18]: Open to suggestions of a different shape for this. Another possibility could be `type ProtectedResourceScopes = Record<string, string[]>`, which would support:
```javascript
["http://applicationD.com", [
  "all.scope",
  {"GET": ["read.scope"]}
]]
```

Comment (jasonnutter) [reply lib/msal-angular/src/msal.interceptor.config.ts:18]: I'll think about this overnight, but I think what you have here is good, as it allows us to easily add more things, and the typings will be more helpful.

File: lib/msal-angular/docs/v2-docs/msal-interceptor.md [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+            ["https://myapplication.com/unprotected", null]
+            ["https://myapplication.com", ["custom.scope"]]
+        ]),
+    }
+    ```
+
+Other things to note regarding the `protectedResourceMap`:
+
+* **Wildcards**: `protectedResourceMap` supports using `*` for wildcards. When using wildcards, if multiple matching entries are found in the `protectedResourceMap`, the first match found will be used (based on the order of the `protectedResourceMap`). 
+* **Relative paths**: If there are relative resource paths in your application, you may need to provide the relative path in the `protectedResourceMap`. This also applies to issues that may arise with ngx-translate. Relative paths should not have a leading slash.
```
Comment (jasonnutter) [inline lib/msal-angular/docs/v2-docs/msal-interceptor.md:147]: "Relative paths should not have a leading slash." Is this always true? I feel like I've seen scenarios where a leading slash was needed.

File: lib/msal-angular/src/msal.interceptor.ts [hunk 3]
```diff
@@ -122,19 +140,62 @@ export class MsalInterceptor implements HttpInterceptor {
...
+            });
+
+            // Only add to all scopes if scopes for endpoint and method is found
+            if (scopesForEndpoint.length > 0) {
+                allMatchedScopes.push(scopesForEndpoint);
             }
+        });
+
+        if (allMatchedScopes.length > 0) {
+            // Returns scopes for first matching endpoint
```
Comment (jasonnutter) [inline lib/msal-angular/src/msal.interceptor.ts:197]: I think it may be worth adding a warning log message here is more than 1 matched scopes are found.

**Review by jasonnutter — APPROVED** [review]
This is great, thanks for the tests and docs!

File: lib/msal-angular/docs/v2-docs/msal-interceptor.md (hunk 2 above)
Comment (tnorling) [reply lib/msal-angular/docs/v2-docs/msal-interceptor.md:147]: Is this a decision we made or a limitation of a dependency? If it's the former I think we should mandate the opposite for consistency (we require a leading slash for redirectUris/postLogoutRedirectUris if you provide a relative url)

File: lib/msal-angular/src/msal.interceptor.config.ts (hunk 1 above)
Comment (jasonnutter) [reply lib/msal-angular/src/msal.interceptor.config.ts:18]: Curious to hear other's thoughts, but I'm cool with the object shape defined here.

File: lib/msal-angular/docs/v2-docs/msal-interceptor.md (hunk 2 above)
Comment (jo-arroyo) [reply lib/msal-angular/docs/v2-docs/msal-interceptor.md:147]: Looking back at customers who have experienced using relative paths, it looks like a leading slash works for some but not for others. We don't know what the relative paths in their apps are going to be, but putting a relative path without a leading slash in the `protectedResourceMap` will work for endpoints with or without a leading slash. 
(i.e. `relative` in the PRM will work for `/relative` and `relative` endpoints, but `/relative` in the PRM will not work for `relative` endpoints).

File: lib/msal-angular/src/msal.interceptor.ts (hunk 3 above)
Comment (jo-arroyo) [reply lib/msal-angular/src/msal.interceptor.ts:200]: Sure. Added.

File: lib/msal-angular/docs/v2-docs/msal-interceptor.md (hunk 2 above)
Comment (jo-arroyo) [reply lib/msal-angular/docs/v2-docs/msal-interceptor.md]: I've updated the guidance on this.

File: lib/msal-angular/src/msal.interceptor.config.ts (hunk 1 above)
Comment (tnorling) [reply lib/msal-angular/src/msal.interceptor.config.ts:18]: I'm a fan of this approach

**Review by tnorling — APPROVED** [review]

**Pull request merged**